CORE-5324 CPI packaging tool should validate the CPB is valid

kyriathar opened the pull request:
Verifies it is a Cpb before packing it into a Cpi.

- adds Cpb verification checks to both `CreateCpi` (V1) and `CreateCpiV2` cli tools.
- enhances testing keystore to include certificates needed for Cpb signatures verification.
- replaces custom Cpb creation/ signing functions in testing with `TestCpbV1Builder` and `TestCpbV2Builder` apis.

josephzunigadaly commented on tools/plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpi.kt:
```diff
@@ -71,26 +76,45 @@ class CreateCpi : Runnable {
+            .verify()
+        true
+    } catch (e: Exception) {
+        System.err.println("Error verifying cpb: ${e.message}")
+        false
```
The program should exit with a non-zero exit code on error. I imagine throwing the exception would do that but also output a stack trace when we want this simple error message. A call to exit(1) after the println might be the best choice.

Calling exit here would save us from having the if statement wrapping all the code too. Hopefully the diff reduces down to just the call to this method.

kyriathar replied:
Good point, will address.

kyriathar replied:
Done -let the exception reach picocli code to be handled by the framework-.

josephzunigadaly commented on tools/plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/signing/VerificationHelpers.kt:
```diff
@@ -0,0 +1,20 @@
+import java.security.cert.X509Certificate
+import net.corda.cli.plugins.packaging.FileHelpers
+
+object VerificationHelpers {
```
Names like `VerificationHelpers` and `SigningHelpers` are open ended. There is a lot that could go in there. It becomes inviting when a future developer is looking for a place to put new code. Over time, classes with names like these build up unrelated code.

If we give this a more specific name, we will reduce the chances of this accumulating new code.

As it stands, this class is reading certificates from a java keystore. So we could call it `JavaKeyStore`, `KeyStoreReader`, `CertificateReader` or `CertificateLoader`. It doesn't have to be one of those, feel free to think up a different name.

The aim is to capture the one (and only one) thing that this class does.

kyriathar replied:
Renamed to `CertificateLoader`.

josephzunigadaly commented on tools/plugins/package/src/test/kotlin/net/corda/cli/plugins/packaging/TestUtils.kt:
```diff
@@ -15,6 +17,25 @@ import org.junit.jupiter.api.Assertions.assertNotNull
 
 internal object TestUtils {
 
+    private val KEY_STORE_PASSWORD = "keystore password".toCharArray()
+
+    private fun privateKeyEntry(alias: String, keyStoreInputStream: InputStream): KeyStore.PrivateKeyEntry {
+        val keyStore = KeyStore.getInstance("PKCS12")
+        keyStoreInputStream.use { keyStoreData -> keyStore.load(keyStoreData, KEY_STORE_PASSWORD) }
+        return keyStore.getEntry(alias, KeyStore.PasswordProtection(KEY_STORE_PASSWORD)) as KeyStore.PrivateKeyEntry
+    }
+
+    private fun resourceInputStream(filePath: String): InputStream {
+        return this::class.java.getResourceAsStream(filePath)
+            ?: throw IllegalStateException("$filePath not found!")
+    }
+
+    val SIGNING_KEY_1_ALIAS = "signing key 1"
+    val SIGNING_KEY_2_ALIAS = "signing key 2"
+
+    val SIGNING_KEY_1 = privateKeyEntry(SIGNING_KEY_1_ALIAS, resourceInputStream("/signingkeys.pfx"))
+    val SIGNING_KEY_2 = privateKeyEntry(SIGNING_KEY_2_ALIAS, resourceInputStream("/signingkeys.pfx"))
+
```
This is an example of how names attract new code. 

These additions represent signing keys, we could move them to new class called `SigningKeys` or `TestSigningKeys`.

kyriathar replied:
Done

josephzunigadaly commented on tools/plugins/package/src/test/kotlin/net/corda/cli/plugins/packaging/CreateCpiTest.kt:
```diff
@@ -127,20 +88,15 @@ class CreateCpiTest {
             )
-        }
-
-        // Expect output to be silent on success
-        assertEquals("", errText)
```
Why was this removed?

kyriathar replied:
Putting those back

josephzunigadaly commented on tools/plugins/package/src/test/kotlin/net/corda/cli/plugins/packaging/CreateCpiTest.kt:
```diff
@@ -105,20 +71,15 @@ class CreateCpiTest {
-            )
-        }
-
-        // Expect output to be silent on success
-        assertEquals("", errText)
```
Another one here

josephzunigadaly commented on tools/plugins/package/src/test/resources/generate-keystore.sh:
```diff
@@ -0,0 +1,41 @@
+        -keyalg ${KEY_ALG} -validity 4000 -keystore signingkeys.pfx -storetype pkcs12 -storepass "${STOREPASS}" \
+        -ext BasicConstraints:critical -ext KeyUsage=cRLSign,digitalSignature,keyCertSign
+
+keytool -genkeypair -alias "${SIGNING_KEY_1}" -dname "cn=CPI Plugin Example - Signing Key 1, o=R3, c=GB" \
+        -keyalg ${KEY_ALG} -validity 4000 -keystore signingkeys.pfx -storetype pkcs12 -storepass "${STOREPASS}"
+
+keytool -genkeypair -alias "${SIGNING_KEY_2}" -dname "cn=CPI Plugin Example - Signing Key 2, o=R3, c=GB" \
+        -keyalg ${KEY_ALG} -validity 4000 -keystore signingkeys.pfx -storetype pkcs12 -storepass "${STOREPASS}" \
```
We only need these two keys. We don't need to build a chain for this testing.

josephzunigadaly replied:
Actually it is closer to real usage with the chain. Forget I said this. No need to change anything for this one.

josephzunigadaly commented on tools/plugins/package/src/test/resources/generate-keystore.sh:
```diff
@@ -0,0 +1,41 @@
+cat rootca.pem key2.pem > key2-chain.pem
+keytool -importcert -alias "${SIGNING_KEY_2}" -file key2-chain.pem -noprompt -keystore signingkeys.pfx -storepass "${STOREPASS}"
+
+# To trust freetsa.org, download their CA cert and import into keystore
+keytool -importcert -alias freetsa -keystore signingkeys.pfx -storepass "${STOREPASS}" -file ~/Downloads/cacert.pem
```
We should remove this. It is not needed for the tests, relies on ~/Downloads and trusts an external service.

kyriathar replied:
Good catch, done.

josephzunigadaly commented on tools/plugins/package/src/test/resources/generate-keystore.sh:
```diff
@@ -0,0 +1,41 @@
+# To trust freetsa.org, download their CA cert and import into keystore
+keytool -importcert -alias freetsa -keystore signingkeys.pfx -storepass "${STOREPASS}" -file ~/Downloads/cacert.pem
+
+rm *.pem
```
This assumes all `.pem` files were generated by this script. It could be dangerous. We should list specific pem files we know the script generates or remove the line entirely.

kyriathar replied:
Done

kyriathar-r3 committed:
CORE-5324 CPI packaging tool should validate the CPB is valid (#1815)

Verifies it is a Cpb before packing it into a Cpi.

- adds Cpb verification checks to both `CreateCpi` (V1) and `CreateCpiV2` cli tools.
- enhances testing keystore to include certificates needed for Cpb signatures verification.
- replaces custom Cpb creation/ signing functions in testing with `TestCpbV1Builder` and `TestCpbV2Builder` apis.
